### PR TITLE
Airflow connections

### PR DESCRIPTION
**Pull request by mgill25**
Basic airflow connections working.

### CODE REVIEW

**Comment by cadicallegari** [comment]
To avoid some issues like last week, I would also suggest adding a test to check duplicated dag names =)

**Comment by mgill25** [comment]
> LGTM
> 
> In the future, maybe we could only pass to the task what it needs.

Sorry I didn't get this. Could you explain?

**Comment by cadicallegari** [comment]
> > LGTM
> > In the future, maybe we could only pass to the task what it needs.
> 
> Sorry I didn't get this. Could you explain?

Instead of reading all the secrets every time that function will be called

https://github.com/fromAtoB/airflow-orchestration/pull/1/files#diff-1ca667d7f4710c943f236ff5bb772fd2R34

the secret keys could be a parameter, then each dag/task could only specify the secrets that it will use

**Comment by mgill25** [comment]
> > > LGTM
> > > In the future, maybe we could only pass to the task what it needs.
> > 
> > 
> > Sorry I didn't get this. Could you explain?
> 
> Instead of reading all the secrets every time that function will be called
> 
> https://github.com/fromAtoB/airflow-orchestration/pull/1/files#diff-1ca667d7f4710c943f236ff5bb772fd2R34
> 
> the secret keys could be a parameter, then each dag/task could only specify the secrets that it will use

Done. :)

**Comment by mgill25** [comment]
> To avoid some issues like last week, I would also suggest adding a test to check duplicated dag names =)

This is also done

File: Dockerfile [hunk 1]
```diff
@@ -11,4 +11,12 @@ RUN pip install --user -r /app/requirements.txt
 # USER root
 # RUN mv sops /usr/local/bin/sops
 # RUN chmod +x /usr/local/bin/sops
+
+# USER root
```
Comment (cadicallegari) [inline Dockerfile]: Maybe we can remove the commented lines =)

File: build/cloud/ship.yaml [hunk 2]
```diff
@@ -1,8 +1,10 @@
 steps:
-- name: gcr.io/cloud-builders/gsutil
-  id: sync-dags-prod
-  # args: [ '-m', 'rsync', '-r', 'dags', 'gs://${_PROD_BUCKET_NAME}/dags' ]
-  args: [ '-m', 'rsync', '-d', '-r', 'dags', 'gs://${_PROD_BUCKET_NAME}/dags' ]
+ - name: docker/compose
+   args: [ 'run', '${_SERVICE_NAME}', 'python', '--version' ]
```
Comment (cadicallegari) [inline build/cloud/ship.yaml]: This step still needed? maybe we can re run the tests here before `deploy` it =)

File: dags/bi_dag.py [hunk 3]
```diff
@@ -6,39 +6,36 @@
...
-kms_keyring = "projects/a2b-exp/locations/global/keyRings/sops/cryptoKeys/sops-key"
-# secrets_filepath = "/app/secrets.enc.yaml"
-secrets_filepath = "/home/airflow/gcs/dags/dagtools/secrets.yaml"
-
-
 dag = DAG(
     'bi_dag',
     default_args=default_args,
     schedule_interval=datetime.timedelta(days=1),
+    on_failure_callback=dagtools.slack_notify_failed
```
Comment (cadicallegari) [inline dags/bi_dag.py]: Nice =)
do we have the slack hook configured in cloud composer?

Comment (mgill25) [reply dags/bi_dag.py]: Not right now. We should do it :)

File: build/cloud/ship.yaml (hunk 2 above)
Comment (mgill25) [reply build/cloud/ship.yaml]: Sure. Do you think we need tests in both scripts?

File: dags/bi_dag.py [hunk 4]
```diff
@@ -6,42 +6,54 @@
...
+
+    start = LatestOnlyOperator(task_id='latest_only')
+
+    mod_list = ['adwords', 'appleads', 'db_external', 'emarsys', 'flix', 'google_analytics', 'facebook', 'bing']
+    dag_list = []
+    for module in mod_list:
+        secret_dict = dagtools.load_secrets_from_connections(keys=auth_key_dict[module])
+        dag_list.append(gen_k8s_dag(image, module, secret_dict))
+
+    start >> dag_list
```
Comment (cadicallegari) [inline dags/bi_dag.py]: Does the DAG need any teardown task?

We can also add a `finished with success` task as the final dependence to send a message to a Slack channel informing that the DAG was succeed

`start >> dag_list >> inform_success_execution >> tear_down`

File: docker-compose.yml [hunk 5]
```diff
@@ -23,12 +23,12 @@ services:
         volumes:
             - ./dags:/usr/local/airflow/dags
             - ./plugins:/usr/local/airflow/plugins
-            - ./tests:/tests
-            - ./hack:/hack
+            # - ./tests:/tests
```
Comment (cadicallegari) [inline docker-compose.yml]: Could it be deleted?

File: hack/setup.sh [hunk 6]
```diff
@@ -14,6 +14,14 @@ airflow connections -a \
...
     --conn_password "T041E87SH/B08K2PN2F/someslack_password"
 
+airflow connections -a --conn_type http --conn_id flix_auth_key --conn_password "foo"
+airflow connections -a --conn_type http --conn_id flix_auth_value --conn_password "foo"
+airflow connections -a --conn_type http --conn_id emarsys_auth_key --conn_password "foo"
+airflow connections -a --conn_type http --conn_id emarsys_auth_value --conn_password "foo"
+airflow connections -a --conn_type http --conn_id fb_app_id --conn_password "foo"
+airflow connections -a --conn_type http --conn_id fb_access_token --conn_password "foo"
+airflow connections -a --conn_type http --conn_id fb_app_secret --conn_password "foo"
+
```
Comment (cadicallegari) [inline hack/setup.sh]: My idea to “productionize” this setup, is to make this script runnable in a given cloud composer environment.
Basically, in another project, we can pass some parameters for the script, the script will then read the secrets (socs can be used here) and apply to the cloud composer environment =)

Comment (mgill25) [reply hack/setup.sh]: Can you explain a bit more. I don't think I get this point?

File: dags/bi_dag.py (hunk 4 above)
Comment (mgill25) [reply dags/bi_dag.py]: What should be a good `tear_down` operator?

Comment (cadicallegari) [reply dags/bi_dag.py]: It depends on the dag. Could be delete delete temporary files or shutdown a cluster for example

File: hack/setup.sh (hunk 6 above)
Comment (cadicallegari) [reply hack/setup.sh]: The same way we can create the connections and variables easily for dev env, we can have this also for prod env, but way more secure :)
But it is good to have, nothing urgent :)

Comment (mgill25) [reply hack/setup.sh]: Yeah, for now I think this is okay. We can do all that as part of subsequent improvement :)

File: dags/bi_dag.py (hunk 4 above)
Comment (mgill25) [reply dags/bi_dag.py]: Do you think this is a necessity? I am thinking these can become future optimizations. For now, our jobs are working and finishing in time, failures are being notified. I think that should be enough to consider it "v1" :)

Comment (cadicallegari) [reply dags/bi_dag.py]: my last comments were about to "productionize".
As far as I know, it is working already so to reduce the lead team we can think about it later =)

The most important part to do next IMO, is to automate the creation of the cloud composer env and setup it later as discussed here https://github.com/fromAtoB/airflow-orchestration/pull/1#discussion_r378503149

the dag we can improve it organically =)

**Comment by cadicallegari** [comment]
Is this PR is not getting too big? =P

**Comment by mgill25** [comment]
> Is this PR is not getting too big? =P

I agree. Lets merge it? :D

**Comment by cadicallegari** [comment]
I approved it when it was only changing connections =P
Maybe we could split it in more concise PRs, but I'm ok with merging it if we don't let huge PRs happen in the future =)

**Pull request merged**